jetty-server/src/main/java/org/eclipse/jetty/server/AbstractLockedHttpInput.java

```diff
@@ -0,0 +1,236 @@
+        {
+            super.setInterceptor(interceptor);
+        }
+    }
+
+    @Override
+    public void addInterceptor(Interceptor interceptor)
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            super.addInterceptor(interceptor);
+        }
+    }
+
+    @Override
+    public void asyncReadProduce()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            super.asyncReadProduce();
+        }
+    }
+
+    @Override
+    public boolean addContent(Content content)
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.addContent(content);
+        }
+    }
+
+    @Override
+    public boolean hasContent()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.hasContent();
+        }
+    }
+
+    @Override
+    public long getContentLength()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.getContentLength();
+        }
+    }
+
+    @Override
+    public boolean earlyEOF()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.earlyEOF();
+        }
+    }
+
+    @Override
+    public boolean eof()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.eof();
+        }
+    }
+
+    @Override
+    public boolean consumeAll()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.consumeAll();
+        }
+    }
+
+    @Override
+    public boolean isError()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.isError();
+        }
+    }
+
+    @Override
+    public boolean isAsync()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.isAsync();
+        }
+    }
+
+    @Override
+    public boolean onIdleTimeout(Throwable x)
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.onIdleTimeout(x);
+        }
+    }
+
+    @Override
+    public boolean failed(Throwable x)
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.failed(x);
+        }
+    }
+
+    /* ServletInputStream */
+
+    @Override
+    public boolean isFinished()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.isFinished();
+        }
+    }
+
+    @Override
+    public boolean isReady()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            return super.isReady();
+        }
+    }
+
+    @Override
+    public void setReadListener(ReadListener readListener)
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            super.setReadListener(readListener);
+        }
+    }
+
+    @Override
+    public int read() throws IOException
+    {
+        try (AutoLock lock = _contentLock.lock())
```

Comment:
I fail to see the reason for different behaviour between HTTP/1 and HTTP/2 with respect to locking?
Why is it OK for 2 applications threads in HTTP/1 to enter `read()` and content on the internal data structures of raw buffers and producers without any memory barriers etc.  yet we mutually exclude when the same application is run on HTTP/2.

It feels like we've not really sorted out what the thread safety of this class should be.  Either we are responsible for thread safety of input or we are not - it should not depend on protocol used... even if only as a side effect of the implementation.

Reply:
This code assumes that there will only ever be a single thread entering `read()` or any other API-exposed method.

The reason why HTTP 1 does without locking is that if the above assumption is true, the whole logic can work without locking (at least in my limited analysis), while in the HTTP 2 case, even if the above assumption is true, we still need locking to avoid internal data races.

Figuring out the exact thread safety requirements of this class is going to be a serious headache. At the moment, my opinion would be to lock all methods in all cases (both HTTP 1 and 2) to have a more common behavior as well as to be on the safe side and work on the thread safety problem in a subsequent refactoring exercise, as just cleaning up the HTTP 1 and 2 logic as well as getting rid of the internal queuing already proves to be quite challenging.

Reply:
I agree that h1 and h2 locking should be the same.
But I don't think we should give up on working out a correct locking model which may need less locking than a default of lock every method.

I think we can safely say that two application threads entering read() at the same time is not supported.  So all we need do is to define how our own callbacks interact with app threads calling HttpInput.  That should be doable... because if we don't understand then we will get it wrong one way or another!

Reply:
More musings.....
Currently the code is written with the assumption that `addContent` can be called spontaneously at any time (for the h2 case), hence that call makes scheduling decisions and significant state changes within `HttpChannelState` and it trusts that the caller will respect the boolean return.

Perhaps in the refactor, calling `addContent` can no longer happen spontaneously?  Perhaps it can only be called once `HttpInput` has decided to progress due to other scheduling decisions?  If so then perhaps it should not be making any scheduling decision nor returning a boolean?  Instead perhaps it just throws an ISE if it is not called as expected?

Reply:
Even more musings.....
During the hangout, I made the suggestion that we should try to always have the thread calling read be the one to do the parsing/producing etc.   However this is not really possible for async, even in HTTP/1, because we have to parse the filled buffer before we know if we have content or not, and only then do we go on to call the onDataAvailable callback.   So that in itself in unachievable.... but we can still strive to make h1 and h2 work in the same way.

Reply:
Sync vs async behavior is another pain point this refactoring made apparent. And yet again, my prescription doesn't vary: investigate to completely understand why the model was made this way and eventually modify it.

Lots of fun ahead! :-)